Feature: Add Triac Phase Control Plugin (_P184_Triac.ino)

thalesmaoa opened the pull request:
This PR introduces a new plugin, _P184_Triac.ino, designed for AC phase control (dimming).

This plugin uses a zero-crossing detection (ZCD) signal to perform its function. It monitors a specific GPIO for this signal, which is typically provided by an external ZCD circuit (e.g., one using an H11AA1 or similar optocoupler).

Based on the timing of the zero-cross signal, the plugin manages the precise firing of a Triac on a separate output GPIO.

This functionality is the fundamental building block for AC phase control, enabling applications such as:
 - AC light dimmers (leading-edge).
 - AC motor speed control.
 - Heating element control (with resistive loads).
 - Synchronizing actions with the AC power line frequency (50/60 Hz).

#### Control Methods

The plugin can be controlled in two different ways via commands, making it flexible for integration into control loops.

 1. Trigger (Phase Angle Control): `triac,trigger,<value>`
    - This command directly controls the firing delay within the AC half-wave.
    - The `value` is a percentage from 0 to 100.
    - This control is inverted:
       - `0` = 100% Power (Triac triggers immediately at the start of the wave).
       - `100` = 0% Power (Triac never triggers).
       - `50` = Triggers at the halfway point of the cycle.
 2. Power (Linearized Power Control): `triac,power,[value]`
    - This command attempts to provide a linearized power output.
    - The `value` is a percentage from 0 to 100.
    - This control is direct:
       - `0` = 0% Power.
       - `100` = 100% Power.
     - The plugin calculates the required trigger delay to achieve the desired power output. This calculation assumes a purely resistive load.

#### Commands and Usage

The plugin can be easily controlled via ESPEasy rules or other controllers (like MQTT or HTTP).

Assuming the plugin's Task Name is `triac`:

```
// Set control using the inverted "trigger" (phase angle) method
triac,trigger,50   // Triggers at 50% of the half-wave
triac,trigger,0    // Triggers at 0% (full power)

// Set control using the "power" (linearized) method
triac,power,70   // Sets output to 70% power (for resistive loads)
triac,power,100  // Sets output to 100% power
``` 

This command-based interface allows the plugin to be easily coupled to any external or internal control loop (e.g., a PID controller, a web UI slider, or an MQTT topic).

thalesmaoa committed:
Triac plugin - Detect zero cross and trigger a triac.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+#ifdef USES_P184
+#define PLUGIN_184
+#define PLUGIN_ID_184     184           // plugin id
+#define PLUGIN_NAME_184   "Triac" // "Plugin Name" is what will be dislpayed in the selection list
```
The name convention is to use a category, a dash (-) and the plugin name, so this could be `Output - Triac`.

(The Output category seems most fitting for this plugin)

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+};
+
+P184_data_struct P184_data;
+
```
Defining this instance globally prevents having multiple concurrent instances of this plugin. Please have a look at how most other plugins have this done, f.e. P140 (a simple plugin example 😃)

TD-er replied:
And/or have a look at the P003 Pulse Counter plugin, as you need to act on GPIO interrupts. This is a special case where you need to also register a pointer to go along with the callback function to store the runtime data.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      dev.PullUpOption       = false;                            // Allow to set internal pull-up resistors.
+      dev.InverseLogicOption = false;                            // Allow to invert the boolean state (e.g. a switch)
+      dev.FormulaOption      = false;                            // Allow to enter a formula to convert values during read. (not possible with Custom enabled)
+      dev.Custom             = false;
```
All Device flags are explicitly initialized to false, so no need to have cpu cycles spent on setting them again. (also a few below)

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      // string += serialHelper_getSerialTypeLabel(event);
+      success = true;
+      break;
+    }
```
Unused / inactive cases can be removed.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      event->Par1 = 2; //getValueCountFromSensorType(static_cast<Sensor_VType>(PCONFIG(P184_OUTPUT_TYPE_INDEX)));
+      success     = true;
+      break;
+    }
```
This case is useful when dynamically changing the number of device values, that's not used in this plugin, so this code can be removed.

thalesmaoa replied:
To be honest. I can't understand how to use this field.

tonhuisman replied:
For this plugin, the marked code can be removed, as you're not changing the number of Values based on some configuration setting.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      event->sensorType = Sensor_VType::SENSOR_TYPE_DIMMER; //static_cast<Sensor_VType>(PCONFIG(event->idx));
+      success           = true;
+      break;
+    }
```
Ditto.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+
+      P184_data.trigger_value = P184_TRIGGER_CONFIG();
+      // Recalculate power value based on the new trigger value from the form
+      for (int i = 0; i <= 100; ++i) {
```
Magic number used here, better would be to use:
`constexpr uint8_t power_to_trigger_lut_size = NR_ELEMENTS(power_to_trigger_lut);`
(just below the `power_to_trigger_lut` array, and the `101` size in that array definition can also be removed. The size is determined at compile-time, this way)
and here, use:
`for (uint8_t i = 0; i < power_to_trigger_lut_size; ++i) {`

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      P184_data.trigger_value = P184_TRIGGER_CONFIG();
+      // Calculate initial power value based on the loaded trigger value
+      for (int i = 0; i <= 100; ++i) {
+        if (pgm_read_byte(&power_to_trigger_lut[i]) <= P184_data.trigger_value) {
```
Ditto.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      if (equals(tmpString, F("triac"))) {
+        String subcmd = parseString(string, 2);
+        String valueStr = parseString(string, 3);
+        long value = valueStr.toInt();
```
Value is already available in `event->Par2`, no need to re-do a string to int conversion.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+    //   success = true;
+    // }
+
+    // case PLUGIN_TEN_PER_SECOND:
```
Commented code can be removed

tonhuisman commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -1495,6 +1495,7 @@ To create/register a plugin, you have to :
 #endif
 
+#define USES_P184
 
```
This will enable the plugin in all builds, unconditionally. That's not correct, it should probably go in the Collection H build (that will be available soon, for now put it in Collection G), the Energy collection and the MAX build definition.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+#define P184_TRIGGER_PIN()         PIN(1)
+
+// int8_t P184_TRIGGER_PIN();
+// uint8_t p184_trigger;
```
More commented code.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+void IRAM_ATTR P184_zero_crossing() {
+  // This is only necessary for debounce
+  // detachInterrupt(P184_data.zero_crossing_pin);
+  if (P184_data.trigger_pin != -1 && P184_data.p184_timer != NULL) {
```
For an example of how instance-independent interrupt handlers can be implemented, please have a look at plugin P008. By moving this code to the plugin_struct sources, you will also avoid compilation warnings about the `iram` section being ignored...

thalesmaoa replied:
Moving it, make sense use `attachInterruptArg`?

TD-er replied:
Yep, if you need to keep some kind of state as static variable, then it makes sense to have a pointer attached to the interrupt, so you can access those variables without creating some elaborate structures to keep track of states per task.

See P003 and P098 for some examples on how to use this `attachInterruptArg` to implement for use with multiple instances of the same plugin.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      P184_FREQ_CONFIG() = getFormItemInt(F("freq"));
+
+      P184_data.trigger_value = P184_TRIGGER_CONFIG();
+      // Recalculate power value based on the new trigger value from the form
```
The plugin is stopped before the PLUGIN_WEBFORM_SAVE function is called, so no need to update settings here, that should (only) be done in PLUGIN_INIT.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+      {
+        pinMode(P184_data.trigger_pin, OUTPUT);
+        digitalWrite(P184_data.trigger_pin, LOW);
+        P184_data.p184_timer = timerBegin(1000000);
```
Magic number?

thalesmaoanz replied:
lol
https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/api/timer.html#timerbegin
I set to 1MHz. Thus I can use the int values in microseconds.

tonhuisman commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+            timerAlarm(P184_data.p184_timer, P184_data.time_us, true, 0);
+            if (loglevelActiveFor(LOG_LEVEL_INFO)) {
+              String log = strformat(F("P184 CMD : Trigger %d%% . Power %d%%"), P184_data.trigger_value, P184_data.power_value);
+              addLogMove(LOG_LEVEL_INFO, log);
```
This can be a single statement, no need to instantiate a new String, only to pass it on.
Then you can also use `addLog` instead of `addLogMove`.

tonhuisman commented:
@thalesmaoa Thanks for this plugin!

I've added a few comments 😅 but that's all positive 👍 
Some work to be done, but it's looking nice already.

There's no reference to hardware or schematics, but that can be addressed in the documentation 😉

tonhuisman commented:
In your example you use this command: `[TaskName],trigger,[value]`, but that's not correct, the taskname prefix is optional, the command always starts with `triac`, and value is required, so the command description could look like this:
`[TaskName.]triac,trigger,<value>`
- Taskname is separated from triac by a dot, and in square brackets, indicating it's optional, and intended to address a second or more instance of the plugin (current code doesn't support that yet). To add some confusion, the taskname can be wrapped in square brackets when used 😆 (this feature is globally available for over 3 years already).
- `<value>` is in range 0..100.

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+  // detachInterrupt(P184_data.zero_crossing_pin);
+  if (P184_data.trigger_pin != -1 && P184_data.p184_timer != NULL) {
+    if (P184_data.trigger_value == 0) {
+      digitalWrite(P184_data.trigger_pin, HIGH);
```
Please look at the code for `DIRECT_pinWrite_ISR` as that's way faster than `digitalWrite`.
And with "way faster" I mean about a factor 100x - 1000x on ESP32-xx

TD-er commented:
Looking at the code and the description, it seems like you turn the Triac 'on' with some delay after the zero-crossing and not on immediately and 'off' after some delay.

I think this will cause way more EMI noise and will introduce way higher peak currents compared to the other way around.

I would expect 'starting' at zero-crossing and turning 'off' after some delay will be way better for your device.
Only thing to really look into is when switching some inductive device as those may generate really high voltages when suddenly having their current interrupted.

tonhuisman commented:
This could probably be implemented by adding a setting to use either the current way or starting at zero-crossing (IIRC this was a 'thing' many years ago, when household lighting dimmers became popular)

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+  uint8_t  power_value      = 0;
+  uint8_t  dead_zone        = 0;
+  uint32_t freq_timing_val  = P184_60HZ_HALF_WAVE_TIME_US_ONE_PERCENT;
+  uint64_t time_us          = 0;
```
No idea why this should be a 64 bit int.
Even if the time was in nano-seconds, it would perfectly fit in an `uint32_t` as it is a sub-second timer.

The reason why it is better to use a 32bit value is because the 64-bit values are dealt with in software and thus are better not to be used in interrupt-handler callback functions.
For sure not for callbacks as frequent as these.

thalesmaoanz replied:
> No idea why this should be a 64 bit int.

Me neither. 
https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/api/timer.html#timeralarm

Also, I've tried using 32bit and I got some strange behavior.

TD-er replied:
Maybe because it isn't declared `volatile` (or `std::atomic<uint32_t>` for modern compilers as used for ESP32)

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+
+void IRAM_ATTR P184_timer_handler() {
+  if (P184_data.trigger_pin != -1) {
+    if (P184_data.trigger_value != 100) {
```
Normally it would be good to have these checks for trigger pin and value, but here you can also make sure not to start the timer at all if these values are unusable.
This makes the callback functions even smaller.
Also these callback functions will be linked to the iram by the compiler and that's a really limited resource. So better make those as small as possible or else other unrelated builds may fail due to insufficient iRAM size.

thalesmaoanz replied:
Yes. You are right. Despite of my best efforts, I was having panicked when deleting the plugin.
Not sure if hardware timer and interrupt takes more time than other to execute. My workaround was to add the check for the pin.

Also, the check for the trigger value is due to grid frequency variations. If frequency is a bit lower, it still triggers. Thus, I need to force it not to.

TD-er replied:
Perhaps you explicitly need to disable the pending timer when deleting the task/plugin?

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,380 @@
+void IRAM_ATTR P184_zero_crossing() {
+  // This is only necessary for debounce
+  // detachInterrupt(P184_data.zero_crossing_pin);
+  if (P184_data.trigger_pin != -1 && P184_data.p184_timer != NULL) {
```
Same for this callback function. You don't need to check for those trigger pin and timer pointers, as you should not even attach to an interrupt when those conditions are not met.

thalesmaoanz replied:
Same here. Panicked when deleting the plugin. Not always, but one in four.

thalesmaoanz commented:
> Triac 'on' with some delay after the zero-crossing

yes! From my previous experience, some devices need some dead time after zero-cross. This can also come due to inductive loads.

> 'off' after some delay

It really doesn't matter. A triac only needs a trigger. It will only block again when current reaches zero.


> I think this will cause way more EMI noise and will introduce way higher peak currents compared to the other way around.

Not sure if I follow.

> I would expect 'starting' at zero-crossing and turning 'off' after some delay will be way better for your device.

Can't do for a [triac](https://www.electronics-tutorials.ws/power/triac.html).
<img width="1200" height="676" alt="image" src="https://github.com/user-attachments/assets/c0749e80-de2e-4045-b804-6b071c30e8c7" />
 

> Only thing to really look into is when switching some inductive device as those may generate really high voltages when suddenly having their current interrupted.

Inductive loads will affect next trigger point. It only turn off when current reaches zero.

TD-er commented:
> Inductive loads will affect next trigger point. It only turn off when current reaches zero.

That's what I meant... if you would cut off the current through an inductive load, it will generate a high  (opposite) voltage as it tries to keep the magnetic field it had.
When you let it 'turn off' at the zero crossing, there is hardly any current left (voltage and current will be out-of-phase) so it will cause less issues when you do it like you're doing now.

However with resistive loads, there is a clear difference as the resistance of the load often is temperature dependent.
Meaning the resistance will be (much) lower at lower temperatures.
So if you would turn it on at the peak voltage, the current will be much higher compared to when you start a cold resistive load at the zero crossing. (or shortly after)
This can also be turned into a 'soft-start' using the triac switching you have right now by slowly increasing the 'on' time.

A capacitive load, like those cheap LED bulbs (or nearly any switching power supply without power factor correction) also may draw way too much when turned on at the peak voltage, as a discharged capacitor acts like a short circuit.
Not sure if those will benefit from a 'slow start' or the opposite and get damaged.

thalesmaoa committed:
Name change
Remove global struct
Remove comments and unnecessary code
Add fixed lut size
REG write for fast pin write
Make interrupt static

thalesmaoa commented:
Hi @tonhuisman and @TD-er . Tried to fill up all the requests.

thalesmaoa committed:
Dont remember

tonhuisman commented on platformio_esp32_envs.ini:
```diff
@@ -311,7 +311,11 @@ extends                   = esp32_IRExt_LittleFS_ETH
+extends                   = env:normal_ESP32_IRExt_4M316k_LittleFS_ETH
+board                     = esp32_4M
+build_flags               = ${env:normal_ESP32_IRExt_4M316k_LittleFS_ETH.build_flags}
+                            -DUSES_P184
```
We've dropped the `_LittleFS_ETH` part from _all_ env names, so when rebasing with `mega`, this env might cause issues, but these are easy to fix 😃

thalesmaoanz replied:
I'm sorry, I've pushed to the wrong origin. I was playing a little bit more with it.

<img width="1600" height="900" alt="image" src="https://github.com/user-attachments/assets/f7987580-c0f8-404e-8da4-609f633b4467" />

https://robotdyn.pt.aliexpress.com/store/1950989/pages/all-items.html?productGroupId=40000005016219&spm=a2g0o.store_pc_allItems_or_groupList.pcShopHead_2562584.1_0&storeId=1950989&sortType=bestmatch_sort&shop_sortType=bestmatch_sort

I'm building a board for temperature control.

<img width="641" height="716" alt="image" src="https://github.com/user-attachments/assets/0ca40575-4682-435c-8a1c-61194abecbc0" />

TD-er replied:
Ehh a few remarks....
Looks like you may be using mains voltage on the headers at the bottom of that screenshot?
Those pins are way too close to the low voltage parts of the PCB.
The area under the dimmer circuit should also be isolated, so rather not have a copper pour on the top layer of the PCB, under the intended dimmer circuit.
If you would like to shield it, then you could add copper pour on the bottom side of your own PCB.
What happens if you scratch the green solder mask with some sharp pieces of the dimmer circuit? Then you expose the GND of the ESP to mains voltage. 
The heat sink of the dimmer circuit is also way too close to the Wemos pins and is probably not isolated from mains.

Also you seem to be using a Wemos ESP32 form factor, which then has the WiFi antenna right above the copper pour area and thus will likely have a very sub-optimal WiFi performance, unless you will be using some external WiFi antenna.

thalesmaoa replied:
Hi @TD-er , I really appreciate your feedback. Thanks!

>  Looks like you may be using mains voltage on the headers at the bottom of that screenshot?

Yes. That is correct.

> Those pins are way too close to the low voltage parts of the PCB.

I performed some calculations based on IEC 60664. For 500V, the minimum suggested clearance is 1.5mm, and I’ve allowed for 1.9mm. However, you’ve brought up a good point: I should have included an isolation slot (milling) in that area, which I missed.

> The area under the dimmer circuit should also be isolated, so rather not have a copper pour on the top layer of the PCB, under the intended dimmer circuit.

I understand. I've decided to use a 3D spacer between them for now. The EMI is significant in Triac-based PCBs, and my intention was for the ground plane to act as a shield.

> Also you seem to be using a Wemos ESP32 form factor, which then has the WiFi antenna right above the copper pour area and thus will likely have a very sub-optimal WiFi performance, unless you will be using some external WiFi antenna.

Do you have any specific tips for improving Wi-Fi performance in this layout?

I have already sent this batch to production, but I will definitely incorporate all of your suggestions into version 2.

TD-er replied:
In open air the rule of thumb is 1 mm per 100V.
So for mains voltage you should keep 3 mm.
An encapsulated PCB trace (covered in solder mask) does have better insulation, so that figure of 1.5mm might be correct.... however... ;)
You're switching some load which can also be inductive. This means the voltages can be much much higher. Just looking at the EMI is also a good indicator there might be some inductive behavior.
N.B. you also have a trace quite close to the other pad of the screw terminal.  I doubt that's at a minimal distance of 1.5 mm of the square pad of J1.

Regarding the antenna.
What you can do for testing, or at least proving the effect of the ground plane on the antenna, is using one or more stacked pin headers to lift the Wemos from the board.
If your Wemos board does have an IPEX connector, you may want to consider connecting an external antenna to it.  Make sure the IPEX connector is actually wired as quite often there is a 0 Ohm resistor which needs to be moved (or 90 degree rotated) to either connect the PCB trace antenna or the IPEX connector.

If you have some PCB material left which doesn't have an area with copper on it (at least on one side), you can place it between the triac circuit board and the Wemos. Make sure no copper is on the side facing the triac board.
1.6 mm PCB material can isolate upto either 16 or 60 kV (not sure about the number), so it may prevent sparking between the triac board and the Wemos.

TD-er replied:
Oh and if you let the boards assemble at for example JLCPCB, you can also use the ESP32 modules they have.
I typically use the 16M module with an UFL connector, so I can use the "MAX" builds and can use whatever antenna I like and place it wherever I like.

thalesmaoa replied:
> Oh and if you let the boards assemble at for example JLCPCB, you can also use the ESP32 modules they have. I typically use the 16M module with an UFL connector, so I can use the "MAX" builds and can use whatever antenna I like and place it wherever I like.

Do mean like this:
https://jlcpcb.com/partdetail/EspressifSystems-ESP32_WROOM_32UEN4/C701344

TD-er replied:
Yep, but then the 16M version so you don't have to worry about build sizes
https://jlcpcb.com/partdetail/736354-ESP32_WROOM_32UEN16/C701346

TD-er replied:
Just add a few capacitors close to the module and 2 transistors for toggling GPIO-0 and reset so you can use those Wemos USB to UART boards for flashing. (CH340 chip, micro USB, 6 pin header)
Just make sure not to use the Vcc pin of those boards as the voltage regulator is a bit underdimensioned and by default wired to just forward the 5V which you need to scratch away, etc...
So just add a row of 6 pins in the correct order accessible on the side to allow you to use a programmer clamp with pogo pins to program the board.

II would just add an 1117 as linear voltage regulator as they are no-nonsense and stable. Not the most power-efficient, but they just work.

Check the datasheet to see if you need to pull-up or -down some GPIO pins (e.g. GPIO-0) and double check this: https://espeasy.readthedocs.io/en/latest/Reference/GPIO.html#best-pins-to-use-on-esp32

Make sure to have a good ground from the pads below the module to the GND plane on the other side.

thalesmaoa committed:
Merge branch 'mega' into feature/zero-crossing-dimmer

thalesmaoa committed:
Merge branch 'mega' into feature/zero-crossing-dimmer

uzi18 commented:
@thalesmaoa it should be possible to detect 50/60 Hz mains - count zero crossing per 1s. if more than lets say 110 than it is 60 Hz.
Also without proper counts per second you may warn user about malfunction and disable triac firing.

uzi18 commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+        const int optionsValsFreq[] = { P184_60HZ_HALF_WAVE_TIME_US_ONE_PERCENT, P184_50HZ_HALF_WAVE_TIME_US_ONE_PERCENT };
+        constexpr int optionsCountFreq = NR_ELEMENTS(optionsValsFreq);
+        const FormSelectorOptions freqSelector(optionsCountFreq, optionsFreq, optionsValsFreq);
+        freqSelector.addFormSelector(F("Grid frequency"), F("freq"),  P184_FREQ_CONFIG());
```
@thalesmaoa it should be possible to detect 50/60 Hz mains - count zero crossing per 1s. if more than lets say 110 than it is 60 Hz.
Also without proper counts per second you may warn user about malfunction and disable triac firing.
Warn user on load page about good quality zero cross signal - this feedback is important for stable output power

TD-er committed:
Merge branch 'mega' into feature/zero-crossing-dimmer

TD-er committed:
[Build] Update referred env:normal_ESP32_IRExt_4M316k

TD-er committed:
[Build] Only allow P184 for ESP32 builds

TD-er committed:
[Build] Shorten env to normal_ESP32_IRExt_4M316k_P184

Using the older (longer) env name, the linker will fail due to loo long argument list.

e.g.
```
[.pio/build/normal_ESP32_IRExt_4M316k_LittleFS_ETH_P184/ESP_Easy_mega_20260413_normal_ESP32_IRExt_4M316k_LittleFS_ETH_P184.elf] sh: Argument list too long
```

tonhuisman commented on platformio_esp32_envs.ini:
```diff
@@ -345,10 +346,10 @@ extends                   = esp32_IRExt
                             
-[env:normal_ESP32_IRExt_4M316k_LittleFS_ETH_P184]
-extends                   = env:normal_ESP32_IRExt_4M316k_LittleFS_ETH
+[env:normal_ESP32_IRExt_4M316k_P184]
```
IMHO this env should be marked as `// FIXME To remove before merging`, as this should be seen as a Custom build.

TD-er replied:
Sure but I was just curious whether it would build anyway as it caused the argument length too long error.

I don't see why there should be a separate build for this anyway as it is already present in a number of other builds.
During development it is useful to have, but before we merge it should be removed indeed.

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+  uint32_t freq_timing_val  = P184_60HZ_HALF_WAVE_TIME_US_ONE_PERCENT;
+  hw_timer_t *timer         = NULL;
+
+  // Funções de interrupção como membros estáticos da struct
```
Please use English also in comments

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+    if (p184_data->trigger_value == 0) {
+      REG_WRITE(GPIO_OUT_W1TS_REG, (1 << p184_data->trigger_pin)); // Fast gpio_set_level(HIGH)
+    } else {
+      REG_WRITE(GPIO_OUT_W1TC_REG, (1 << p184_data->trigger_pin)); // Fast gpio_set_level(LOW)
```
We do have a GPIO_Direct_Access lib for this, so it is platform independent as this differs per ESP board (even among ESP32-variants)

Just search for `DIRECT_pinWrite` and `DIRECT_pinRead` and for use in ISR (callback) functions: `DIRECT_pinWrite_ISR` and `DIRECT_pinRead_ISR`

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+  static void IRAM_ATTR timer_handler(void *arg) {
+    P184_data_struct* p184_data = static_cast<P184_data_struct*>(arg);
+    if (p184_data->trigger_value != 100) {
+      REG_WRITE(GPIO_OUT_W1TS_REG, (1 << p184_data->trigger_pin)); // Fast gpio_set_level(HIGH)
```
Ditto, use `DIRECT_pinWrite_ISR`

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+    }
+    case PLUGIN_INIT:
+    {
+      P184_data_struct* p184_data = (P184_data_struct*)getPluginTaskData(event->TaskIndex);
```
When calling `PLUGIN_INIT`, I don't see how `getPluginTaskData` should ever return something other than a `nullptr`.

thalesmaoa replied:
It was for safety. Should I remove it?

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+    {
+      P184_data_struct* p184_data = (P184_data_struct*)getPluginTaskData(event->TaskIndex);
+      if (p184_data == nullptr) {
+          // Aloca a memória se ainda não existir
```
Please use English

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+          // Aloca a memória se ainda não existir
+          p184_data = new P184_data_struct;
+          if (p184_data == nullptr) {
+              // Falha na alocação de memória
```
Please use English

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+      {
+        gpio_set_level(p184_data->trigger_pin, LOW);
+        if (p184_data->timer != NULL) {
+          timerEnd(p184_data->timer);
```
Shouldn't this be ended before doing anything else in the destruction of the `P184_data_struct` ?

Also I think it would make sense to have this code (in the `PLUGIN_EXIT` case) in the destructor of that `P184_data_struct` class.

TD-er commented on src/_P184_Triac.ino:
```diff
@@ -0,0 +1,310 @@
+        p184_data->trigger_pin = GPIO_NUM_NC;
+      }
+      
+      clearPluginTaskData(event->TaskIndex);
```
Have to check, but I think this is also done after the `PLUGIN_EXIT` call is done.

Edit: Yep is done right after the `PLUGIN_EXIT` call

TD-er commented:
@thalesmaoa 
I have made some small changes so the PR could be built again given all other recent changes in the code.
I also added some comments on the code. Please have a look as it would otherwise not even work correctly on other ESP32-types and/or for all usable pins. (e.g. GPIOs with pincount >= 32 will not work right now)

Then the only thing missing I guess is some documentation and it would be nice to have some schematics along with it on how to use it (preferably using optocouplers)

drzorg82 commented:
I tried this plugin, but it didn't work. I built it using this schematic, but I used a MOC3043. The light bulb lights up at full power at any trigger value.
<img width="826" height="380" alt="Снимок экрана 2026-04-15 154619" src="https://github.com/user-attachments/assets/07ec1e9a-5d82-4c41-9c32-3bdaaff3457b" />
<img width="1060" height="906" alt="Снимок экрана 2026-04-15 154803" src="https://github.com/user-attachments/assets/1307850d-0312-43f2-8865-ad205a5b5571" />
<img width="1912" height="889" alt="Снимок экрана 2026-04-15 155326" src="https://github.com/user-attachments/assets/477f3f5c-c838-40ee-b730-a7bde5b2e966" />

TD-er commented:
Just as a simple test, can you check to add a pulse counter task to the interrupt pin (nothing else requesting interrupts on that pin) and see if you can see roughly 50 Hz signal?
I assume you have a pull-up resistor attached to that pin also as the optocoupler will only pull the pin low, not up.

TD-er commented:
Also wouldn't it make sense to have a diode anti-parallel to the LED of the optocoupler? Just to make sure the reverse voltage over the LED is limited.  Usually those only can handle a few Volt.
The 1N4001 should also do this, but I guess you better pick a 1N4007 or something like that as the 1N4001 can only handle 50-ish V reverse Voltage.

TD-er commented:
And I wonder if the plugin is designed to act only on the positive part of the sine and not on the negative?

drzorg82 commented:
I didn't quite understand what you wanted to say, but I did it this way.
.
.
.
<img width="1903" height="379" alt="Снимок экрана 2026-04-16 095701" src="https://github.com/user-attachments/assets/dba628df-68eb-4b7f-9f4d-93e97f2896ea" />
<img width="911" height="875" alt="Снимок экрана 2026-04-16 100058" src="https://github.com/user-attachments/assets/a5149747-ee97-457c-be54-5cfe925428f5" />
.
.
.
and only with these trigger settings the light bulb glows at half power (Trigger Deadzone has no effect)
.
.
.
<img width="851" height="901" alt="Снимок экрана 2026-04-16 095732" src="https://github.com/user-attachments/assets/deecd02d-e1e7-4a16-8261-89b13eede971" />
<img width="1901" height="403" alt="Снимок экрана 2026-04-16 095856" src="https://github.com/user-attachments/assets/10350f13-c2d7-4985-8bc8-8e0969959ada" />
.
.
For example, if I set the trigger to 20, the light starts blinking randomly from half power to full power. Then it might glow at full power, then it starts blinking randomly again. And this happens with any other settings.
And only with the trigger set to 99 does the light constantly glow at half power.
.
.
<img width="1905" height="430" alt="Снимок экрана 2026-04-16 100356" src="https://github.com/user-attachments/assets/6a319316-6bef-4898-961e-aea7ebbb3426" />

I didn't change anything in the diagram.

TD-er commented:
You should **not** have both the Triac and the Pulse Counter tasks active at the same time as those interrupts on the same pin might have an effect on eachother.
But at least it looks like the optocoupler is outputting a pulse which looks plausible assuming the pulsecounter task is set to an interval of 1 sec.

This suggestion was just a test to see if the optocoupler for the trigger input was actually working or not.

Do you have a pull-up resistor present on the trigger pin? I doubt the internal pull-up resistor of the ESP may be enough to create a strong pull, which may cause a delay in the signal.

thalesmaoa commented:
Hi @TD-er! I will try to fix it ASAP.

@drzorg82, there can be any different reasons for that behaviour that you have.



> I tried this plugin, but it didn't work. I built it using this schematic, but I used a MOC3043. The light bulb lights up at full power at any trigger value. <img alt="Снимок экрана 2026-04-15 154619" width="826" height="380" src="https://private-user-images.githubusercontent.com/124128771/578621467-07ec1e9a-5d82-4c41-9c32-3bdaaff3457b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYzNDEwMjEsIm5iZiI6MTc3NjM0MDcyMSwicGF0aCI6Ii8xMjQxMjg3NzEvNTc4NjIxNDY3LTA3ZWMxZTlhLTVkODItNGM0MS05YzMyLTNiZGFhZmYzNDU3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQxNlQxMTU4NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMjZjODU3Zjk0Mzg1MzhjZGQyMTdiYjU4M2RmNjRiMzJlNmY5ZmE0NWVmNGZhNDM1MTU5ODFlYjc5MzIxZDRkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.qhf9U63UBecD-ZWxPEAXQse5DKsgagpw1R5yxnRWUSU"> <img alt="Снимок экрана 2026-04-15 154803" width="1060" height="906" src="https://private-user-images.githubusercontent.com/124128771/578621499-1307850d-0312-43f2-8865-ad205a5b5571.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYzNDEwMjEsIm5iZiI6MTc3NjM0MDcyMSwicGF0aCI6Ii8xMjQxMjg3NzEvNTc4NjIxNDk5LTEzMDc4NTBkLTAzMTItNDNmMi04ODY1LWFkMjA1YTViNTU3MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQxNlQxMTU4NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MTBjODZmNTFhOTcwZjcxOWFhZmQwNTgyOTFkZTYxN2QzODlmNjY4NDY3YWVhNGY5N2M4NWVkMjQ4ZDNmNjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.lRJ5l4y7QFE92Bkn-LA1Rhyhld2kcoF0F04MNMXR_Jc"> <img alt="Снимок экрана 2026-04-15 155326" width="1912" height="889" src="https://private-user-images.githubusercontent.com/124128771/578623278-477f3f5c-c838-40ee-b730-a7bde5b2e966.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYzNDEwMjEsIm5iZiI6MTc3NjM0MDcyMSwicGF0aCI6Ii8xMjQxMjg3NzEvNTc4NjIzMjc4LTQ3N2YzZjVjLWM4MzgtNDBlZS1iNzMwLWE3YmRlNWIyZTk2Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQxNlQxMTU4NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMTIyMTUwMzgxZTA3OWIzYzhlYmM3MDZmYThkM2EzOTI0ZjQwNWQ1YmYxOTYzZjRjZjE3MzkwNjQyNGZjZDM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.r3AWdZ7hXdVLgED1IB-qAp6ywE-x_Cs3vhnpAvkd1Ck">

The proposed circuit works only if you chose (not FALLING, neither RISING) CHANGE option. It trigger on RISING and on FALLING. Be aware that you may have debouncing problems. Specially due to the high inductance load.

TD-er commented:
> The proposed circuit works only if you chose (not FALLING, neither RISING) CHANGE option. It trigger on RISING and on FALLING. Be aware that you may have debouncing problems. Specially due to the high inductance load.

Hmm do you get a duty-cycle of 50% with this, when using an optocoupler acting only on the positive halves of the sine curve?

thalesmaoa commented:
> Hmm do you get a duty-cycle of 50% with this, when using an optocoupler acting only on the positive halves of the sine curve?

you always need to trigger when zero cross. This is one strategy of PLL. Since we only have half-cycle. It must trigger for rising and for falling. The timing will be a little bit different for each case due to dead time band.

My suggestion is to add a counter parallel diode to the opto and remove the series one.

drzorg82 commented:
> Do you have a pull-up resistor present on the trigger pin? I

No. I'm using it according to the schematic diagram above. Later, I'll try adding a pull-up (or down?) resistor and replacing one diode with a diode bridge.

thalesmaoanz commented:
> > Do you have a pull-up resistor present on the trigger pin? I
> 
> No. I'm using it according to the schematic diagram above. Later, I'll try adding a pull-up (or down?) resistor and replacing one diode with a diode bridge.

You need a pull down.

TD-er commented:
> > > Do you have a pull-up resistor present on the trigger pin? I
> > 
> > 
> > No. I'm using it according to the schematic diagram above. Later, I'll try adding a pull-up (or down?) resistor and replacing one diode with a diode bridge.
> 
> You need a pull down.

<img width="498" height="248" alt="image" src="https://github.com/user-attachments/assets/c382379a-2dc8-4ba7-8eb4-85010e0f2162" />


Ehhh? Are you sure?

thalesmaoanz commented:
> > > > Do you have a pull-up resistor present on the trigger pin? I
> > > 
> > > 
> > > No. I'm using it according to the schematic diagram above. Later, I'll try adding a pull-up (or down?) resistor and replacing one diode with a diode bridge.
> > 
> > 
> > You need a pull down.
> 
> <img alt="image" width="498" height="248" src="https://private-user-images.githubusercontent.com/3751318/579542596-c382379a-2dc8-4ba7-8eb4-85010e0f2162.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYzODczODUsIm5iZiI6MTc3NjM4NzA4NSwicGF0aCI6Ii8zNzUxMzE4LzU3OTU0MjU5Ni1jMzgyMzc5YS0yZGM4LTRiYTctOGViNC04NTAxMGUwZjIxNjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTdUMDA1MTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmExYTZiZTU5NDY4MTZkNmIwODMzMTA5NmI3YTExNmVkMjQ3NjkyZWY1YThmNDMyZjE1OWMwN2M0MDUzNjY2NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.MqXoFVtZiIR5SF7qSyOdbrSOzFWU9ROI4DwXkeDfaCk">
> Ehhh? Are you sure?

Sorry, I was talking about the trigger pin to avoid floating conditions. Interrupt pin is a pull up.

thalesmaoa committed:
Merge branch 'mega' into feature/zero-crossing-dimmer

thalesmaoa committed:
Update suggestions

thalesmaoa committed:
Remove test flag

TD-er commented:
Due to lots of other things to do the last week(s), I lost a bit what has to do with this PR.
Maybe one of you can quickly summarize what still needs my attention?  (will also post this on other open PRs)

thalesmaoa commented:
Hi @TD-er. I've replied to all of your comments. I've changed everything in accordance. I've tested the code and it is working fine. Due to the lack of updates from @drzorg82 , I suppose things are working fine.

drzorg82 commented:
I apologize for the silence, but I haven't tried your recommendations yet. Unfortunately, I don't have enough time for my hobby. Now I probably won't be able to try it until fall. But I'll definitely let you know once I try it.

TD-er committed:
Merge branch 'mega' into feature/zero-crossing-dimmer

SuksAE commented:
I just stumbled across this, so I don't know if I am too late for a feature request:

It would be great if you could implement a pulse packet mode.
In this mode, only complete halve sine waves are switched. You switch the triac on at (or a very short time after) zero crossing and keep it on for n% halve sine waves. Then follows a pause of 100%-n half sine waves where the triac is off. 

As phase control is limited to loads of less than 400W (at least in europe) this would enalbe the use of this plugin for larger resistive heaters (like water heating or electric HVAC heaters) where there is no big difference to phase control because of the thermal mass. Furthermore the electromagnetic noise will be reduced to a minimum.

Additionally it would be great to also have a configuration factor (e.g. y) to set the number of halve waves the triac is on or off for each percent of dimming (so 100% would be y*100 sine halve waves, 75% would be 25*y sine halve waves off and 75*y sine halve waves on, ...)

I hope my request makes sense and as mentioned - I hope I am not too late...

Thanks

thalesmaoa commented:
Not sure if I follow your request. 

I've already built a three phase 25kW with the same structure. If I sleep for n-cycles, is the same as solid state relays. PWM with SSR should be enough for if.

Can you explain better?

TD-er commented:
What I understood from the request is that it is better to have X full sine periods on and then Y full sine periods off.
By switching during zero-crossing, you introduce nearly no noise and the grid is being used for the full sine periods.
Turning on/off during a sine will cause noise and extra imbalance on the grid. (or at least local mains net).

And since the heat-capacity of the water is so big, there is no benefit of switching in PWM mode during a sine.

thalesmaoa commented:
A SSR has a zero cross circuit built in. A PWM should be enough to solve it.

SuksAE commented:
> Not sure if I follow your request.
> 
> I've already built a three phase 25kW with the same structure. If I sleep for n-cycles, is the same as solid state relays. PWM with SSR should be enough for if.
> 
> Can you explain better?

You are right - using a zero crossing solid state relais does exactly the same, but at a higher cost (money wise). I fully understand if you want to finish this as is and do not accept any feature requests at this late time...

Thanks anyway.

thalesmaoa commented:
It's not like that @SuksAE . I want to help you. I just can't understand your request.

SuksAE commented:
> What I understood from the request is that it is better to have X full sine periods on and then Y full sine periods off. By switching during zero-crossing, you introduce nearly no noise and the grid is being used for the full sine periods. Turning on/off during a sine will cause noise and extra imbalance on the grid. (or at least local mains net).
> 
> And since the heat-capacity of the water is so big, there is no benefit of switching in PWM mode during a sine.

and this is exactly the reason why phase control is limited to light loads with <400W by law - at least in europe...
but thalesmaoa is right that a SSR can do the job almost the same way as I described it. And I don't know if the 400W limitation is relevant for the rest of the world (I would guess yes, but I don't know for sure).

SuksAE commented:
> It's not like that @SuksAE . I want to help you. I just can't understand your request.

you are right, using a SSR does in fact the same thing. But I am not sure if ESP Easy supports generating a PWM signal <100Hz. I have to look into that...

And by the way - I currently have no imminent need for such mechanism - just about a year ago I had a discussion with a friend about controlling a water heating element using PV excess energy and this would by handy back then - now he uses a much simpler approach but with much less granularity in dimming resolution...
We both did not see the solution with a SSR controlled by a PWM signal - I will forward this idea - maybe he wants to change his system once more to get a better resolution...

Thanks again

TD-er commented:
For longer pulses, you may want to have a look at the `longpulse` and `longpulse_ms` commands.
.. or at least at its implementation :)

SuksAE commented:
> For longer pulses, you may want to have a look at the `longpulse` and `longpulse_ms` commands. .. or at least at its implementation :)

I will - thanks

thalesmaoa commented:
Good to know, however, let me insist that there is a better way to solve your problem. It is uses this exactly plugin.

Heaters are generally a slow system. You can easily deploy a PI using rules (there are some examples). Using the already presented plug-in, you just control the power based on temperature. `triac,power,<from 0 up to 100>`. This is performed with zero cross, which can be used for any power. It smooth the control and avoid overshooting.

This topology avoid pulsing the heater and improve its life. The drawback is that you will start to have grid quality problems as the power increase. For higher powers you may need some snubber and filters to trigger the thyristor, like I did for 25kW.

TD-er commented:
Hmm this triggers an idea which has been lingering in my mind for quite some time...

I always wanted to have some PID control mechanism, but had not yet a good idea on how to implement it in ESPEasy without having it implemented multiple times in several plugins.

Like for a PID you need one or more inputs and an output to create some feedback loop and prevent overshoots and with some tuning of the PID values, you can even let it act as a 'slow start'.

Let me think about this a bit more, but I think we can make this feature request suggestion also work with some PID like approach.

For this the `triac,power` command could also have like a second optional argument, where you can set the time in half sine periods over which to distribute the power duty cycle.
Like default for this 2nd parameter then is `0` (cycle) for PWM within a cycle. With the second parameter > 0, you set the minimum 'on' time in half since cycles.

Then with a PID control (in rules I guess?) you can then control the power parameter.

So I'm thinking about a new 'type' of variable in rules, which you initialize with the PID parameters and some initial state/value.
This way it is easier to keep all required values/states together and you can just 'feed' the PID controller periodically with new inputs like the measured temperature.

Only thing that may make this a bit more complex is what to do when the power command is called mid-dutycycle? Should that one be finished first and the next call using the new values?  When having the 'minimal cycle' value set to a relative high value and the 'power' low (e.g. 1%), the full cycle can take some time.  On the other hand, you only would set those for a rather slow process anyway...

thalesmaoa commented:
That would be nice, but I'd like to share some of my experiences.

What I love about ESPEasy:

 - It's easy and plug-and-play.
 - I can edit configurations right from my phone.
 - It covers 90% of the features I need.

What bothers me about ESPEasy:

For the remaining 10% of my use cases, I need speed. A 1-second delay is too high, and the rules engine introduces too much overhead. I love the concept of rules, but they are just too slow.

Why I prefer to avoid ESPHome:

After years of working with ESP devices, I know I eventually forget how or why I configured something, and I lose track. ESPEasy is strictly standalone, which is its greatest strength. While it would be nice to pull in some external features, keeping it standalone is the best approach.

Final thoughts:

If I were to suggest the next step, I'd look into opening up the scope for fine-tuning. It would be great to replace the standard rules with a faster Virtual Machine or allow for pre-compiled plugins. Right now, it's difficult to understand the execution flow (even with AI assistance) when trying to develop a new plugin or controller.
If we could build our own custom logic—like a PID controller—using raw C code or a compiled .wasm file, it would solve many of these limitations.

TD-er commented:
Well there is someone who was working on increasing rules processing speed significantly by pre-parsing the rules and building a complete tree with parsed rules code.
The main problem why this has not yet been implemented in ESPEasy is that it requires some changes in the rules syntax/grammar which would render existing rules incompatible without some conversion.

However I do see why this would be the way forward as the current rules implementation can only handle upto 30-ish event calls per sec. (which is quite optimistic)

If you need some extra speed for a PID which needs to run at like a few-100 times a second, then I totally get it why it should be part of the plugin. (or maybe called from within the plugin while being set and configured via rules)

Under the hood it is all C++ code, so I see no reason why for speed reasons you can't run some code directly from the plugin code, or maybe even some callback function.
As long as we can keep the amount of code duplication to an absolute minimum, as I have been spending (and likely will spend a lot more) countless hours trying to minimize the build size of ESPEasy. (and minimizing memory requirements)

thalesmaoa commented:
Don't get me wrong, things are perfect. I can't complain at all. After some time using a tool, you want to get the best out of it. No doubt. I'm just trying to share some of my personal experience with ESPEasy, trying to push it to its limits.

I just wanted to share. Here are some examples:

1.  I do have an industrial PID running in some places. ESPEasy is only for collecting data. Since I'm limited to a 1s publish rate over MQTT, UDP, etc., my control bandwidth can't be faster than that. Also, I need to integrate other sensors with other ESPEasy instances. It acts on many different drivers, so I can't deploy the control logic inside the ESP. For this specific reason, I would need a faster sensor response.

 2. Filter implementation. It is better and easier to implement filters inside ESPEasy. For a simple first-order filter, it's fine, but depending on the sensor, it has become too difficult to handle it using rules.

 3. Calibration. ESPEasy has some features for calibration, but it still has some limitations. I had a board reading a 4-20mA pressure sensor. However, with the pressure and the custom venturi I've developed, I am actually measuring flow. I need ADC calibration, then pressure calibration, and after that, flow calibration. I need to access those 3 values and implement the filter. I considered creating a plugin, but I wasn't convinced that made sense.

 4. I2C or Serial. The proposed plugins work. But I wasn't able to use them without building my own new plugin. It was extremely slow.
 
I'm on a new project now where I need to use the HC-12. I'm still struggling with whether I should dev a new controller or just create a simple program. I love ESPEasy and just want to use it for everything.

TD-er commented:
About the response-rate to communicate to other nodes, if it has hard-realtime requirements, then I doubt MQTT is the best solution as the MQTT broker doesn't really have real-time promises on when to notify subscribed clients.
In ESPEasy you can use `looptimer` with msec resolution, but it is limited to about 30x per sec as absolute max.
The only guarantee I can give you with the msec resolution timer is that it will be rescheduled with the given interval from the previous set timer.
So let's assume 100 msec interval and the first call was scheduled with millis() at exactly 10000, then any next scheduled timer will be at 10000 + N* 100. So there will only be jitter in the looptimer, no drift (apart from the inaccuracy of the internal clock of the ESP).
Whenever the intended new scheduled time had already passed, the next interval will be used. This means there isn't even a guaranteed number of loops per sec.
That's also why I highlight rows in the timingstats page which take over 100 msec as those will for sure mess up scheduled timings and need special attention for optimization.

ESPEasy p2p commands are probably faster than MQTT as every controller in ESPEasy has a process queue to not overload the receiving end.  p2p commands are sent immediately.
N.B. you can try to tweak the controller queue settings for MQTT by reducing the minimum send interval. This is set to 100 msec by default, which may even be a bit too short for some setups. For WiFi it may not make much of a difference when setting it to a lower value, simply because of the 102.4 msec beacon interval used by WiFi.  For Ethernet this delay in transactions does not exist.  This is also why loading a webpage via Ethernet typically feels much faster than via WiFi.



Filter implementation.
I think we both mean exactly the same without realizing :)
What I mentioned with a new filter object, like how variables are now used in ESPEasy rules is simply that there would be a C++ struct which keeps all parameters as member and when feeding a new value (e.g. `filterSet,foo,1234` ), it would update its internal state and when referenced (e.g. `[filter#foo]` ) will be replaced by the latest output value.
This can be just about anything, like an IIR, FIR, sliding window, PID, etc.

But I don't see why you shouldn't be able to call those filter objects from within the C++ code of a plugin, if speed is required.
This way you can refer to the filter output value from various points in ESPEasy (e.g. logging, showing on a display, etc) and still update it way more frequent than would be possible via rules.

TD-er commented:
Have you seen the calibration features offered by the internal ADC plugin (P002) ?
I still have plans to separate that code into some separate code which then could be used by any plugin.
But as I mentioned I didn't have it clear yet how to make this generic.

The way I see it, conceptually there isn't that much of a difference between calibration or a filter.
The main difference is that calibration doesn't typically have a timing component (thus no response delay) and is device specific.

Also you may want to apply calibration first and then a filter.

So I was thinking you may want to have some selectors to enable some calibration and/or a filter within a task.
Perhaps even per task-value, like we now also have the formula fields.  However this may become quite cluttered quite fast.
So maybe some formula like reference to a defined filter would be nice to have and then some other place in the web UI to tune/view a filter and/or calibration.

TD-er commented:
> I2C or Serial. The proposed plugins work. But I wasn't able to use them without building my own new plugin. It was extremely slow.

Do you have a link or some more info on this?
